### DISCUSSION TITLE
How to report custom errors ideomatically?

### DISCUSSION THREAD

**Discussion by hmeine** [winnow-rs/winnow#504] [Q&A]
I have been using winnow for a few AoC quizzes and have now been trying to port some more serious binary file format parser from nom to winnow. Mostly, I love how it turns out, but this file format has a few peculiarities, so I need to deviate from the stuff tought in the tutorial and find myself running into errors around the error type:

- I am writing functions that return `PResult<T>`, but how do I now report errors from custom logic? For instance, I have successfully parsed a configuration table that's part of the file header, and I later need to to look up various (string) keys in that table. How should I report that such a key is missing? Given a std rust `Result` with my custom error, I have tried `.map_err(|_| ErrMode::Cut(InputError::new(tag_list_input, ErrorKind::Tag)))?` for instance, but that's both ugly and not working yet (InputError vs. ContextError, I guess).
- Similarly, at one point I read out the size of a section, then use `input.split_at()` to get two inputs for further parsing. Trying to parse the above-mentioned configuration table from the stripped-off section, I have been trying `my_configuration_table_parser.parse(section)` (instead of the `.parse_next()`s usually used), but that also yields the wrong error type (in the sense that I cannot just pass it on using `?`). Of course, I could use `.parse_next()` on a mut ref, but then I would have to check myself that it has parsed the full section.

Maybe I am just revealing that I have not yet understood all best practices w.r.t. Rust's error handling (and all the available generic third-party crates).

**Comment by hmeine** [comment]
I have satisfied the compiler using
```
        .map_err(|e| {
            ErrMode::Cut(ContextError::from_external_error(
                &tag_list_input,
                ErrorKind::Tag,
                e,
            ))
        })?
```
but

1. is this technically correct and intended? And
2. (how) can I hide this? Maybe by defining some `Into` or `From` traits that allow me to just use the `?` in my parsing code?

**Reply by epage** [reply]
That or `try_map` are the main ways of doing it.  `try_map` is what I use in `toml_edit` but I also am doing a "conversion" at the same time.

You could simplify your code to
```rust
        .map_err(|e| {
            ErrMode::from_external_error(
                &tag_list_input,
                ErrorKind::Tag,
                e,
            ).cut()
        })?
```

> (how) can I hid this? Maybe by defining some Into or From traits that allow me to just use the ? in my parsing code?

In what way are you wanting to "hide" this?  Are you more just wanting to make this reusable?  I think that depends on context (especially since it needs `tag_list_input`).

**Reply by hmeine** [reply]
Thanks a lot for your answers. 
I did not find (or even expect) the `ErrMode::from_external_error()` – that's already shorter, and I admit that my question was ignoring the necessity for the `&tag_list_input`. As long as the reference needs to stay, I also do not see a way to further optimize.

**Comment by epage** [comment]
> Similarly, at one point I read out the size of a section, then use input.split_at() to get two inputs for further parsing. Trying to parse the above-mentioned configuration table from the stripped-off section, I have been trying my_configuration_table_parser.parse(section) (instead of the .parse_next()s usually used), but that also yields the wrong error type (in the sense that I cannot just pass it on using ?). Of course, I could use .parse_next() on a mut ref, but then I would have to check myself that it has parsed the full section.

You could do `(first_half, winnow::combinator::eof)`, `(second_half, winnow::combinator::eof)`

**Comment by epage** [comment]
In general, if you are porting from `nom`, it'd be good to contrast with what you are doing with `nom` and why that isn't working as most things should translate fairly well.

**Reply by hmeine** [reply]
Yes, sometimes I succeeded with a more gradual transition, but not particularly often.

There are many differences I have to cope with – for instance, ATM my tests fail because `winnow:::ascii::dec_uint` does not accept leading zeros (while `nom::character::complete::u16` did).
